File: app/src/main/java/dev/leonlatsch/olivia/database/interfaces/ChatInterface.java
package dev.leonlatsch.olivia.database.interfaces;

import com.activeandroid.query.Delete;
import com.activeandroid.query.Select;
import com.activeandroid.query.Set;
import com.activeandroid.query.Update;

import java.util.List;

import dev.leonlatsch.olivia.database.model.Chat;
import dev.leonlatsch.olivia.database.model.Message;

public class ChatInterface extends BaseInterface {

    private static final String QUEUE_CID_WHERE = "cid = ?";
    private static final String QUEUE_MID_WHERE = "mid = ?";
    private static final String QUEUE_UID_WHERE = "uid = ?";

    private static ChatInterface chatInterface; // Singleton

    private ChatInterface() {
    }

    public void saveMessage(Message message) {
        Message savedMessage = new Select().from(Message.class).where(QUEUE_MID_WHERE, message.getMid()).executeSingle();
        if (savedMessage == null) {
            message.save();
        }
    }

    public void saveChat(Chat chat) {
        if (!chatExists(chat)) {
            chat.save();
        }
    }

    public void updateChat(Chat chat) {
        new Update(Chat.class).set("cid = ?, uid = ?, unread_messages = ?, last_message = ?, last_timestamp = ?", chat.getCid(),
                chat.getUid(), chat.getUnreadMessages(), chat.getLastMessage(), chat.getLastTimestamp())
                .where(QUEUE_CID_WHERE, chat.getCid()).execute();
    }

    public void setMessageSent(Message message) {
        new Update(Message.class).set("sent = ?", message.isSent() ? 1 : 0).where(QUEUE_MID_WHERE, message.getMid()).execute();
    }

    public boolean messageExists(Message message) {
        return new Select().from(Message.class).where(QUEUE_MID_WHERE, message.getMid()).executeSingle() != null;
    }

    public boolean chatExists(Chat chat) {
        return new Select().from(Chat.class).where(QUEUE_CID_WHERE, chat.getCid()).executeSingle() != null;
    }

    public void deleteChat(String cid) {
        new Delete().from(Chat.class).where(QUEUE_CID_WHERE, cid).execute();
        new Delete().from(Message.class).where(QUEUE_CID_WHERE, cid).execute();
    }

    public void deleteAll() {
        new Delete().from(Chat.class).execute();
        new Delete().from(Message.class).execute();
    }

    public List<Chat> getALl() {
        return new Select().from(Chat.class).execute();
    }

    public List<Message> getAllUnsentMessages() {
        return new Select().from(Message.class).where("sent = ?", false).execute();
    }

    public Chat getChat(String cid) {
        return new Select().from(Chat.class).where(QUEUE_CID_WHERE, cid).executeSingle();
    }

    public Chat getChatFromMessage(Message message) {
        return new Select().from(Chat.class).where(QUEUE_UID_WHERE, message.getFrom()).executeSingle();
    }

    public Chat getChatForContact(String uid) {
        return new Select().from(Chat.class).where(QUEUE_UID_WHERE, uid).executeSingle();
    }

    public List<Message> getMessagesForChat(String cid) {
        return new Select().from(Message.class).where(QUEUE_CID_WHERE, cid).execute();
    }

    public Message getMessage(String mid) {
        return new Select().from(Message.class).where(QUEUE_MID_WHERE, mid).executeSingle();
    }

    public static ChatInterface getInstance() {
        if (chatInterface == null) {
            chatInterface = new ChatInterface();
        }

        return chatInterface;
    }
}

File: app/src/main/java/dev/leonlatsch/kolibri/database/interfaces/ChatInterface.kt
package dev.leonlatsch.kolibri.database.interfaces

import com.activeandroid.query.Delete
import com.activeandroid.query.Select
import com.activeandroid.query.Update

import dev.leonlatsch.kolibri.database.model.Chat
import dev.leonlatsch.kolibri.database.model.Message

/**
 * Database interface to persist chats and messages
 *
 * @author Leon Latsch
 * @since 1.0.0
 */
object ChatInterface {

    private const val QUEUE_CID_WHERE = "cid = ?"
    private const val QUEUE_MID_WHERE = "mid = ?"
    private const val QUEUE_UID_WHERE = "uid = ?"

    val all: List<Chat> = Select().from(Chat::class.java).execute()

    val allUnsentMessages: List<Message> = Select().from(Message::class.java).where("sent = ?", false).execute()

    fun saveMessage(message: Message) {
        val savedMessage = Select().from(Message::class.java).where(QUEUE_MID_WHERE, message.mid).executeSingle<Message>()
        if (savedMessage == null) {
            message.save()
        }
    }

    fun saveChat(chat: Chat) {
        if (!chatExists(chat)) {
            chat.save()
        }
    }

    fun updateChat(chat: Chat) {
        Update(Chat::class.java).set("cid = ?, uid = ?, unread_messages = ?, last_message = ?, last_timestamp = ?", chat.cid,
                chat.uid, chat.unreadMessages, chat.lastMessage, chat.lastTimestamp)
                .where(QUEUE_CID_WHERE, chat.cid).execute()
    }

    fun setMessageSent(message: Message) {
        Update(Message::class.java).set("sent = ?", if (message.isSent) 1 else 0).where(QUEUE_MID_WHERE, message.mid).execute()
    }

    fun messageExists(message: Message): Boolean =
            Select().from(Message::class.java).where(QUEUE_MID_WHERE, message.mid).executeSingle<Message>() != null

    fun chatExists(chat: Chat): Boolean =
            Select().from(Chat::class.java).where(QUEUE_CID_WHERE, chat.cid).executeSingle<Chat>() != null

    fun deleteChat(cid: String) {
        Delete().from(Chat::class.java).where(QUEUE_CID_WHERE, cid).execute<Chat>()
        Delete().from(Message::class.java).where(QUEUE_CID_WHERE, cid).execute<Chat>()
    }

    fun deleteAll() {
        Delete().from(Chat::class.java).execute<Chat>()
        Delete().from(Message::class.java).execute<Message>()
    }

    fun getChat(cid: String): Chat = Select().from(Chat::class.java).where(QUEUE_CID_WHERE, cid).executeSingle()

    fun getChatFromMessage(message: Message): Chat {
        return Select().from(Chat::class.java).where(QUEUE_UID_WHERE, message.from).executeSingle()
    }

    fun getChatForContact(uid: String): Chat {
        return Select().from(Chat::class.java).where(QUEUE_UID_WHERE, uid).executeSingle()
    }

    fun getMessagesForChat(cid: String): List<Message> {
        return Select().from(Message::class.java).where(QUEUE_CID_WHERE, cid).execute()
    }

    fun getMessage(mid: String): Message {
        return Select().from(Message::class.java).where(QUEUE_MID_WHERE, mid).executeSingle()
    }
}
